torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,234 @@
+    def _is_quantized_linear(model: torch.nn.Module, fqn: str):
+        return hasattr(model, "scale")
+
+    @torch.no_grad()
+    def create_qlinear(linear):
+        def _get_qinfo(linear):
+            # qdq_weight shape: (oc, ic)
+            qdq_weight = linear.weight.clone()
+            device = qdq_weight.device
+            # scales, zeros shape: (oc, n_groups)
+            scales = linear.scale.to(device)
+            zeros = linear.zp.to(device)
+
+            # Requantize the qdqweight to get the int_data
+            orig_shape = qdq_weight.shape
+            oc, ic = orig_shape
+            groupsize = linear.group_size
+            assert (
+                ic % groupsize == 0
+            ), f"expect k % groupsize == 0, but got {ic % groupsize}"
+            n_groups = ic // groupsize
+
+            # Check the shapes of scales and zeros with int_data
+            scales_zeros_expected_shape = torch.Size([oc, n_groups])
+            assert (
+                scales.shape == scales_zeros_expected_shape
+            ), f"expect scales shape {scales_zeros_expected_shape}, but got {scales.shape}"
+
+            assert (
+                zeros.shape == scales_zeros_expected_shape
+            ), f"expect zeros shape {scales_zeros_expected_shape}, but got {zeros.shape}"
+
+            flatten_scales = scales.reshape(-1, 1)
+            flatten_zeros = zeros.reshape(-1, 1)
+            gs_shape = (-1, groupsize)
+            int_data = (
+                qdq_weight.reshape(gs_shape)
+                .div(flatten_scales)
+                .add(flatten_zeros)
+                .round()
+                .reshape(orig_shape)
+                .to(torch.int32)
+            )
+
+            # Shift the zeros to align with tiny gemm.
+            # The dequantization process in tiny gemm:
+            #   tiny_dequant = (tinny_quant - 8) * scale + tinny_zp
+            # The dequantization porcess in auto-round
+            #   dequant = (quant - zp) * scale
+            # To align with tiny gemm:
+            #   dequant = (quant - 8 + 8 - zp) * scale
+            #           = (quant - 8) * scale + (8 - zp) * scale
+            #              \___/                \______________/
+            #            tiny_quant                 tiny_zp
+            zeros = (8 - zeros) * scales
+
+            # Pack to tinygemm reqiured format
+            # Hard code inner_k_tiles = 2
+            inner_k_tiles = 2
+
+            packed_int_data = torch.ops.aten._convert_weight_to_int4pack(
+                int_data, inner_k_tiles
+            )
+            scales_and_zeros = ao_quant.utils.pack_tinygemm_scales_and_zeros(
+                scales.to(torch.bfloat16), zeros.to(torch.bfloat16)
+            )
+            return packed_int_data, scales_and_zeros
+
+        int_data, scales_and_zeros = _get_qinfo(linear)
+
+        woq_weight = ao_quant.Int4WeightOnlyQuantizedLinearWeight(
```

Comment:
this is what we are planning to deprecate, please use `AffineQuantizedTensor` instead

Reply:
maybe I should remove the alternative module swap path, I'd recommend to use https://github.com/pytorch/ao/blob/00b76c4291cdb6c5563f83936341afb9bac39f25/tutorials/calibration_flow/static_quant.py#L46-L65 for this use case so that you can use the same int4 code path that will be maintained by torchao

Reply:
Thanks, update it to use `AffineQuantizedTensor`  at [e5ffcca](https://github.com/pytorch/ao/pull/581/commits/e5ffcca6e5633f9751305ddbdd985033a6c52306).